Add support for hiding products in the Point of Sale

malinajirka opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

This PR adds a new POS (Point of Sale) product visibility taxonomy to WooCommerce core, allowing merchants to control    
  which products are visible in Point of Sale applications. The taxonomy will be eventually deleted/replaced when a proper implementation for robust location-inventory-management is added to WooCommerce core.

Key changes:

  1. **New taxonomy registration** - Registers `pos_product_visibility` taxonomy for products with a `pos-hidden` term to mark products as hidden from POS.
  2. **Admin UI** - Adds a "Available for POS" checkbox in the product editor's Advanced tab.
  3. **REST API support** - Adds a `pos_products_only` boolean parameter to the Products and Variations REST API endpoints for filtering.
  4. **POS catalog integration** - Updates the POS catalog feed to automatically exclude products and variations marked as hidden.
                                                                                                                   
  This feature enables merchants to hide products for in-person sales, 
  useful for:                                                                                                
  - Products only available for online purchase (e.g., large size items, ...)                                  
  - Products that require special handling not suitable for POS.
  - Wholesale-only or retail-only product differentiation.

Note: The "Show in Point of Sale" checkbox is currently displayed for all product types, we'll limit it to products supported in the POS in [this follow up PR](https://github.com/woocommerce/woocommerce/pull/62667). Feel free to test all the changes in one go in that PR.

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

Thank you @iamgabrielma for collaborating/pair-programming on these changes!

Closes #WOOMOB-1894

### Screenshots or screen recordings:

<!-- If this PR includes UI changes, please provide screenshots or a screen recording for clarity. -->
<!-- This section can be removed if not applicable. -->

| Before | After |
| ------ | ----- |
| <img width="1004" height="362" alt="Screenshot 2026-01-05 at 17 53 35" src="https://github.com/user-attachments/assets/d51d96ea-bffa-4a92-a1e9-45aa4199c8ab" /> | <img width="1008" height="405" alt="Screenshot 2026-01-06 at 9 05 25" src="https://github.com/user-attachments/assets/e8d62cf0-cf94-4ade-940a-6eddf10c64a4" /> |


<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:
                                           

 #### Test 0: Admin UI - Disable POS
                                                                                                                   
  1. WooCommerce -> Settings -> Advanced -> Features -> disable Point of Sale
  2. Go to Products > All Products                                                                                 
  3. Edit any existing product (or create a new simple product)                                                    
  4. Navigate to the Product data > Advanced tab                                                                   
  5. Expected: You should NOT see a "Show in Point of Sale" checkbox                                                   
  6. WooCommerce -> Settings -> Advanced -> Features -> enable Point of Sale

                                                                                                                   
 #### Test 1: Admin UI - Show/Hide product in POS                                                                      
                                                                                                                   
  1. Go to Products > All Products                                                                                 
  2. Edit any existing product (or create a new simple product)                                                    
  3. Navigate to the Product data > Advanced tab                                                                   
  4. Expected: You should see a "Show in Point of Sale" checkbox                                                   
  5. Verify the checkbox is checked by default for existing products                                               
  7. Uncheck the "Show in Point of Sale" checkbox                                                                  
  8. Click Update to save the product                                                                              
  9. Refresh the page                                                                                              
  10. Expected: The checkbox should remain unchecked
                                       
                                                                                                                                          
#### Test 2: REST API Products Endpoint - Filter by POS visibility                                                                      
                                                                                                                   
  1. Create at least 2 products:                                                                                   
    - Product A: "Show in Point of Sale" checked (visible)                                                         
    - Product B: "Show in Point of Sale" unchecked (hidden)                                                        
  2. Make a REST API request to get only POS-visible products:                                                     
  curl -X GET "https://yoursite.com/wp-json/wc/v3/products?pos_products_only=true" \                                     
    -u consumer_key:consumer_secret                                                                                
  3. Expected: Only Product A should be returned                                                                   
  4. Make a REST API request with pos_products_only set to false:                                                      
  curl -X GET "https://yoursite.com/wp-json/wc/v3/products?pos_products_only=false" \                                    
    -u consumer_key:consumer_secret                                                                                
  5. Expected: Both products should be returned (no filtering)
  6. Make a REST API request without the pos_products_only parameter:                                                    
  curl -X GET "https://yoursite.com/wp-json/wc/v3/products" \                                                      
    -u consumer_key:consumer_secret                                                                                
  7. Expected: Both products should be returned (no filtering)         


#### Test 3: REST API Variations Endpoint - Filter by POS visibility                                                                      
                                                                                                                   
  1. Create at least 2 variable products with some variations:                                                                                   
    - Product A: "Show in Point of Sale" checked (visible)                                                         
    - Product B: "Show in Point of Sale" unchecked (hidden)                                                        
  2. Make a REST API request to get only POS-visible variations:                                                     
  curl -X GET "https://yoursite.com/wp-json/wc/v3/variations?pos_products_only=true" \                                     
    -u consumer_key:consumer_secret                                                                                
  3. Expected: Only variations of Product A should be returned                                                                   
  4. Make a REST API request with pos_products_only set to false:                                                      
  curl -X GET "https://yoursite.com/wp-json/wc/v3/variations?pos_products_only=false" \                                    
    -u consumer_key:consumer_secret                                                                                
  5. Expected: All variations should be returned (no filtering)
  6. Make a REST API request without the pos_products_only parameter:                                                    
  curl -X GET "https://yoursite.com/wp-json/wc/v3/variations" \                                                      
    -u consumer_key:consumer_secret                                                                                
  7. Expected: All variations should be returned (no filtering)                                                   
                                                                                                                   
    #### Test 4: POS catalog feed excludes hidden products                                                                
                                                                                                                   
  1. Ensure you have 
  - one simple product set as visible in POS
  - one variable product set as hidden in POS with at least 1 variation
  - one variable product set as visible in POS with at least 1 variation                                     
  3. Trigger the POS catalog generation endpoint "https://yoursite.com/wp-json/wc/pos/v1/catalog/create?_product_fields=id&_variation_fields=id,parent_id"                               
  4. Expected: Products with "Show in Point of Sale" unchecked should NOT appear in the generated catalog file (variations of a hidden product should NOT appear in the file either).         
                                                                                                                   
                                    
                                                                                                                   
      #### Test 5: Frontend store unaffected                                                                                
                                                                                                                   
  1. Hide a product from POS (uncheck "Show in Point of Sale")                                                     
  2. Visit the product on the frontend store                                                                       
  3. Expected: The product should still be visible and purchasable on the frontend                                 
  5. Add the product to cart and complete a test order                                                             
  6. Expected: The order should process normally - POS visibility should not affect online sales  


<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

The above

### Milestone

<!-- DO NOT remove or modify this section (other than to check the box). -->

<!-- milestone-target-selection -->
- [x] Automatically assign milestone for the **[next WooCommerce version](../blob/trunk/plugins/woocommerce/woocommerce.php#L6)**
<!-- /milestone-target-selection -->

> **Note:** Check the box above to have the milestone automatically assigned when merged.
> Alternatively (e.g. for point releases), manually assign the appropriate milestone.


### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

malinajirka committed:
Add taxonomy indicating whether product should be hidden in POS

malinajirka committed:
Add pos visibility to product detail

malinajirka committed:
Move pos visibility from product overview to advanced

malinajirka committed:
Filter out products hidden in pos from generated catalog

malinajirka committed:
Merge branch 'trunk' into issue/woomob-1894-add-pos-visibility-taxonomy-to-core

malinajirka commented on plugins/woocommerce/includes/admin/meta-boxes/class-wc-meta-box-product-data.php:
```diff
@@ -427,6 +427,18 @@ public static function save( $post_id, $post ) {
+		$pos_visible = isset( $_POST['_pos_visible'] ) && 'yes' === $_POST['_pos_visible'];
+		if ( $pos_visible ) {
+			wp_remove_object_terms( $post_id, 'pos-hidden', 'pos_product_visibility' );
+		} else {
```
📓 AI added
```
if ( ! term_exists( 'pos-hidden', 'pos_product_visibility' ) ) {
				wp_insert_term( 'pos-hidden', 'pos_product_visibility' );
			}
```
Just to double check, it's not needed, right?

malinajirka committed:
Remove todo

malinajirka committed:
Hide POS Taxonomy settings when POS flag disabled

malinajirka committed:
Remove support for variations in POS product visibility taxonomy

malinajirka committed:
Make pos product visibility taxonomy private

malinajirka committed:
Add tests for changes in /products endpoint

malinajirka committed:
Rename pos_visible to visible_in_pos for clarity

malinajirka commented on plugins/woocommerce/src/Internal/ProductFeed/Integrations/POSCatalog/POSIntegration.php:
```diff
@@ -54,7 +54,15 @@ public function get_id(): string {
 		return array(
-			'type' => array( 'simple', 'variable', 'variation' ),
+			'type'      => array( 'simple', 'variable', 'variation' ),
+			'tax_query' => array(
```
📓 We need to verify performance of this change on some large store!

malinajirka replied:
I've performed a naive test on a store with 18k products and it was as fast as the previous solution - I didn't spot a measurable difference.

RadoslavGeorgiev replied:
Querying by taxonomies is not the most performant thing in the world. If the test store had thousands of terms (categories, tags, etc.), you might feel a bit of a performance downgrade.

However, it is the better solution here. There is way less downside in comparison to post meta (the initial plan).

malinajirka committed:
Fix linting errors

malinajirka committed:
Add changelog entry - add ability to hide products in POS

malinajirka committed:
Suppress lint warning about slow query

malinajirka committed:
Use product_object->get_id() to get product id

malinajirka committed:
Update phpstan baseline

malinajirka committed:
Escape visible_in_pos query param

malinajirka commented on plugins/woocommerce/includes/class-wc-post-types.php:
```diff
@@ -212,6 +212,35 @@ public static function register_taxonomies() {
+			 * @since 9.8.0
+			 * @param array $post_types Array of post types.
+			 */
+			apply_filters( 'woocommerce_taxonomy_objects_pos_product_visibility', array( 'product' ) ),
```
This is a valid point - I'm not sure what is the best approach here. It feels a bit off to add the taxonomy to all variations - especially, since they can be created later on and potentially not even through WPAdmin. I'm inclined towards doing the filtering on the endpoints level, wdyt?

RadoslavGeorgiev replied:
This is more of a business decision, rather than a dev one. The main question is whether one would even want to control variation visibility though their parent products.

Technically, you could listen to variable product updates, and add/remove the tag for variations. However, I don't truly think that it is necessary.

However, I think the `FeedValidator` for POS should check the taxonomies of the variable product, and add an error for variations that should be hidden.

malinajirka replied:
From a business perspective, we'd like to avoid displaying variations that have a hidden parent product, however, we don't need to control visibility of each variation separately.

This might be actually quite tricky. Even if we update the `FeedValidator` the `/variations` REST endpoint will continue returning these variations - the POS uses this endpoint for incremental updates. We technically won't ever display them in the POS, since the only way to access variations **at the moment** is through parent product. However, if we add support for variation-search down the road, we might end up in quite a pickle.

cc @iamgabrielma?

malinajirka replied:
> However, I think the FeedValidator for POS should check the taxonomies of the variable product, and add an error for variations that should be hidden.

How do you imagine this error would work/look like? This might be a solution even to the problem I shared in the above comment. If the catalog endpoint indicates which variations should be hidden, we can continue hiding them even when we fetch an updated version from `/variations` endpoint.

iamgabrielma replied:
> We technically won't ever display them in the POS, since the only way to access variations **at the moment** is through parent product.
> This might be actually quite tricky [...] if we add support for variation-search down the road, we might end up in quite a pickle.

IMHO I think supporting variation-search down the road would come with changes to the way we handle products as a whole, right? Since as you mention, right now is not possible to handle variations if the parent product is not eligible for POS. We've worked with this constraint since the inception of POS, so while I see the point of allowing granular handling of variations and allow/disallow them in POS per-variation basis (this variation is available in this POS, while this other isn't), I think this should emerge naturally later on from actual product needs, at the moment I'd say the expectation is that if the parent is not eligible, its variations aren't either.

If we decide to add variation-search down the road, this will come with other changes as well which can be evaluated along, ie: Perhaps when the time comes this feature has already been removed in favor of distribution channels.

My understanding is that, at the moment the only downside would be that we still reach to the variations endpoint for incremental updates, despite not showing them in the UI, right?

malinajirka replied:
> I see the point of allowing granular handling of variations and allow/disallow them in POS per-variation basis (this variation is available in this POS, while this other isn't), I think this should emerge naturally later on from actual product needs, at the moment I'd say the expectation is that if the parent is not eligible, its variations aren't either.

Just to ensure we are on the same page -> I'm not suggesting to introduce per-variation handling - I agree it needs to come from product needs and when that's needed, we'll take care of it.

> If we decide to add variation-search down the road, this will come with other changes as well which can be evaluated along, ie: Perhaps when the time comes this feature has already been removed in favor of distribution channels.

This is the tricky part - there is like 99% chance, that folks implementing search won't even realize the API provides variations that are actually not supposed to be visible in POS. They'll just implement a local FTS in the database. The POS will end up showing variations in search that should not be available in POS.

We might be the ones implementing it so we might remember, but IMO we should try to avoid introducing traps for future devs or future ourselves.

iamgabrielma replied:
> This is the tricky part - there is like 99% chance, that folks implementing search won't even realize the API provides variations that are actually not supposed to be visible in POS. They'll just implement a local FTS in the database. The POS will end up showing variations in search that should not be available in POS.

Gotcha! Is there any straight-forward way to make these inherit or reach for the visibility of the parent? Or are we entering in the mud with this as well? (just writing it down it feels is going to be trouble for the future). Perhaps some sort of internal/deprecation warning (not sure how core handles these) would suffice to be mindful when building on top of it?

malinajirka replied:
I can see the following options:
1. As Radoslav suggested, have a hook that will add/remove the taxonomy to all variations of the product when the product visibility updates. We'd also need to copy parent visibility when a new variation is created.
2. Ignore it as you suggested and hope that we'll implement a proper inventory/location handling which will cover this case. It's a trap for future devs (especially the ones implementing POS variation search), but in the end the UX isn't terrible - the worst case scenario, we display variation in POS that should be hidden. It's a merchant facing software not customer facing, so it might not be such a big deal after all.
3. Introduce `visible_in_pos` flag even for /variations endpoint and check parent's visibility. Not sure if there is a performant way to do this, but feels like the cleanest and easy solution. We'd also introduce the same logic for the catalog endpoint (FeedValidator).
4. Update just the FeedValidator so it returns these variations but with some extra flag. We'd then need to handle unsupported variations on the client - we'd also need to make sure not to override this flag when fetching incremental updates from /variations endpoint. Feasible but feels ugly.

I personally like 3 the most, but after thinking about this more I'm starting to agree that 2 might be good enough.

RadoslavGeorgiev replied:
Do you want to give this patch a try?

```diff
diff --git a/plugins/woocommerce/src/Internal/ProductFeed/Integrations/POSCatalog/FeedValidator.php b/plugins/woocommerce/src/Internal/ProductFeed/Integrations/POSCatalog/FeedValidator.php
index da116097ae..d94ff47b91 100644
--- a/plugins/woocommerce/src/Internal/ProductFeed/Integrations/POSCatalog/FeedValidator.php
+++ b/plugins/woocommerce/src/Internal/ProductFeed/Integrations/POSCatalog/FeedValidator.php
@@ -9,6 +9,7 @@ declare(strict_types=1);
 
 namespace Automattic\WooCommerce\Internal\ProductFeed\Integrations\POSCatalog;
 
+use Automattic\WooCommerce\Enums\ProductType;
 use Automattic\WooCommerce\Internal\ProductFeed\Feed\FeedValidatorInterface;
 
 if ( ! defined( 'ABSPATH' ) ) {
@@ -29,6 +30,13 @@ final class FeedValidator implements FeedValidatorInterface {
 	 * @return array Array of validation issues.
 	 */
 	public function validate_entry( array $entry, \WC_Product $product ): array { //phpcs:ignore VariableAnalysis
+		if ( ProductType::VARIATION === $product->get_type() ) {
+			if ( has_term( 'pos-hidden', 'pos_product_visibility', $product->get_parent_id() ) ) {
+				return array(
+					'Parent product is hidden from the POS catalog',
+				);
+			}
+		}
 		return array();
 	}
 }
```

We will still have to loop over all variations in PHP, but this should prevent the variation from appearing in the catalog.

The `has_term()` call should be fairly performant, as terms are cached while loading posts (products).

malinajirka replied:
Thanks for the patch!

This will work for the catalog endpoint, however, it might get more tricky for the /variation endpoint. If we do filtering like this, the `X-WP-Total, X-WP-TotalPages` headers will be off and pagination will be a bit weird. But perhaps, improving at least the /catalog endpoint might be good enough. 

There will still be a scenario in which POS might display a hidden variation, but it'll be more rare - since the variation will be displayed only if it's modified between full-catalog-syncs.

I've pushed the patch in 9da9d093415cc12ac1335aff31a79179bcaa0962. I'll test it later today.

malinajirka replied:
Just to close the loop, due to some other issues ([comment](https://github.com/woocommerce/woocommerce/pull/62534#discussion_r2669374371)) I decided to enable the taxonomy for variations and add the term to all variations of a hidden product.

21cd41b7729afb4eb13558a38412a4e334f1e9df

iamgabrielma commented:
👋 I gave it a quick go, but the API tests didn't seem to work on my end:
- ✅ Disabling POS from advanced settings hide the checkbox, re-enabling it works normally and shows the checkbox.
- ✅ Once installed, all products I checked had the option "Show in Point of Sale" enabled by default.
- ✅ Setting checkbox to false and saving the product persists the checkbox state. If I disable POS and re-enable it, state persists.
- ❌ Enabling/disabling the checkbox on a product seems to return the same response no matter if `pos_visible=true/false`. For reference, I get the same amount of products for either case:
```
## false
curl -i \
"https://indiemelon.mystagingwebsite.com/wp-json/wc/v3/products?consumer_key=KEY&consumer_secret=SECRET&pos_visible=false"

## true
curl -i \
"https://indiemelon.mystagingwebsite.com/wp-json/wc/v3/products?consumer_key=KEY&consumer_secret=SECRET&pos_visible=true"

## no param
curl -i \
"https://indiemelon.mystagingwebsite.com/wp-json/wc/v3/products?consumer_key=KEY&consumer_secret=SECRET"

## response
x-wp-total: 86
x-wp-totalpages: 9
```

Version is `10.5.0-dev` shared through slack

malinajirka committed:
Update text for visible in POS toggle

malinajirka commented:
Thank you for testing @iamgabrielma! 

>  the API tests didn't seem to work on my end:

I found the issue :facepalm: sorry... I renamed the parameter to visible_in_pos and I forgot to update the description.

> Version is 10.5.0-dev shared through slack. Unsure if related, but upon installing I got this error (everything seems to work normally though):

AFAICT with my limited understanding, this is not related to the changes.

malinajirka committed:
Hide label description behind the tip icon

malinajirka committed:
Do not change pos visibility when field not present

malinajirka commented on plugins/woocommerce/src/Internal/ProductFeed/Integrations/POSCatalog/POSIntegration.php:
```diff
@@ -54,7 +54,16 @@ public function get_id(): string {
+			// phpcs:ignore WordPress.DB.SlowDBQuery.slow_db_query_tax_query
+			'tax_query' => array(
+				array(
+					'taxonomy' => 'pos_product_visibility',
```
💬 I'm still wondering if we should do this filtering only when a certain query parameter is present to simplify potential future migration away from this taxonomy. Wdyt?

RadoslavGeorgiev replied:
Those are sort of base args, and should not be dynamic. This is more than fine for now. Even if you switch to a different technique to store visibility, the catalog would always ignore hidden products, right?

malinajirka replied:
It will likely become more than just a visibility boolean flag. For example, a product might be hidden in New York but visible in San Francisco. Wdyt @RadoslavGeorgiev?

RadoslavGeorgiev replied:
Let's think about this when we get there. I imagine that POS visibility and store selection would be two separate things. No need to dig further into this at the moment.

malinajirka commented on plugins/woocommerce/phpstan-baseline.neon:
```diff
@@ -7581,7 +7581,7 @@ parameters:
 			message: '#^Variable \$product_object might not be defined\.$#'
 			identifier: variable.undefined
-			count: 3
+			count: 4
```
What is the usual approach here? I'm using a variable that is used in other instances in the file, so feels a bit weird that something extra needs to be done - that's why I opted for updating the baseline. 🤔

RadoslavGeorgiev commented on plugins/woocommerce/includes/class-wc-post-types.php:
```diff
@@ -212,6 +212,35 @@ public static function register_taxonomies() {
+			/**
+			 * Filter the post types that the POS product visibility taxonomy is attached to.
+			 *
+			 * @since 9.8.0
```
The next WooCommerce version is `10.5.0`, and this change will likely be relesed with it. The easiest place to find it is to look at the auto-assigned milestone in GitHub (`10.5.0` in this case). Those tags should reflect it.

malinajirka replied:
I would have sworn I fixed this already 🤷 - anyway, fixed in ab85807ca8fd0ca11ce7337e4389aacb34d5e775.

RadoslavGeorgiev commented on plugins/woocommerce/includes/class-wc-post-types.php:
```diff
@@ -212,6 +212,35 @@ public static function register_taxonomies() {
+					'hierarchical'      => false,
+					'show_ui'           => false,
+					'show_in_nav_menus' => false,
+					'query_var'         => is_admin(),
```
You can simply make this `false`. The query var is used when queries are constructed based on the current URL, and that will never happen for POS products. We're constructing our own queries.

malinajirka replied:
Updated in ebfd1ad9d7e66e8fdf951b4f78e6bc51928ffe24

RadoslavGeorgiev commented on plugins/woocommerce/includes/rest-api/Controllers/Version3/class-wc-rest-products-controller.php:
```diff
@@ -1947,6 +1954,13 @@ public function get_collection_params() {
+			'type'              => 'boolean',
+			'sanitize_callback' => 'wc_string_to_bool',
+			'validate_callback' => 'rest_validate_request_arg',
+		);
```
Ah, the classic situation with boolean parameters :D

I get what providing `true` here would do, but what would providing `false` do? Would it skip products that are visible in POS, or it would not filter? I know the answer, but the parameter name is kind of weird.

Given that this will only be relesed with/after the POS catalog endpoint, do you need to add the parameter here?

malinajirka replied:
We will still use this endpoint for incremental updates, so yes I believe we need to add this parameter.

> I get what providing true here would do, but what would providing false do? Would it skip products that are visible in POS, or it would not filter? I know the answer, but the parameter name is kind of weird.

Do you have some tips on how to name it better? I fully agree with your thoughts, I just couldn't come up with a better name. I guess we could change the logic and when it's false, return all products and name it `pos_products_only`. Wdyt @RadoslavGeorgiev ?

RadoslavGeorgiev replied:
`pos_products_only` is 👌

malinajirka replied:
Updated in fe1639c67fdf9b223b00ca8b1bbfc657083eac2b.

RadoslavGeorgiev commented on plugins/woocommerce/includes/admin/meta-boxes/class-wc-meta-box-product-data.php:
```diff
@@ -427,6 +427,16 @@ public static function save( $post_id, $post ) {
+			} else {
+				wp_set_object_terms( $post_id, 'pos-hidden', 'pos_product_visibility' );
+			}
+		}
```
Actually, checkboxes do not necessarily end up as strings in PHP. If the checkbox is checked, it's `value` property appears as a value in the `$_POST` array. If it is not checked, there is no value in `$_POST['_visible_in_pos']`.


```suggestion
        $visible_in_pos = isset( $_POST['_visible_in_pos'] ) && 'yes' === wc_clean( wp_unslash( $_POST['_visible_in_pos'] ) );
        if ( $visible_in_pos ) {
            wp_remove_object_terms( $post_id, 'pos-hidden', 'pos_product_visibility' );
        } else {
            wp_set_object_terms( $post_id, 'pos-hidden', 'pos_product_visibility' );
        }
```

malinajirka replied:
@RadoslavGeorgiev Hmm interesting, the code you are suggesting is what I originally had. However, then I realized that we will be hiding the checkbox for unsupported products in [this follow up PR](https://github.com/woocommerce/woocommerce/pull/62667). I was worried that when the checkbox is hidden, we might unintentionally add the `pos-hidden` taxonomy to the product even though we shouldn't. How would you suggest to go about this? Should we somehow hide the checkbox for unsupported products, but still send its current value with form submission?

RadoslavGeorgiev replied:
I don't think you should worry about the visiblity of the checkbox. As long as it is on the page, it's in the DOM. This means that whatever value is set, it will remain the one that will be saved.

malinajirka committed:
Fix version in @since comment

malinajirka committed:
Update query_var for pos product visibility taxonomy

malinajirka commented:
Thanks so much for the review @RadoslavGeorgiev! I've followed up on your comments. Looking forward to your thoughts, thanks!

malinajirka committed:
Filter out variations with hidden parent from catalog feed

malinajirka committed:
Rename visible_in_pos to pos_products_only and simplify behavior

The visible_in_pos parameter was confusing because setting it to false
returned only hidden products rather than all products. The new
pos_products_only parameter has clearer semantics:

- pos_products_only=true: return only POS-visible products
- pos_products_only=false or omitted: return all products (no filtering)

This makes the API more intuitive since false now means "don't apply
this filter" rather than "show hidden products only".

malinajirka committed:
Update plugins/woocommerce/includes/admin/meta-boxes/class-wc-meta-box-product-data.php

Co-authored-by: Radoslav Georgiev <rageorgiev@gmail.com>

malinajirka committed:
Remove misleading outdated comment

malinajirka committed:
Fix lint/formatting in class-wc-meta-box-product-data.php

prettyboymp commented on plugins/woocommerce/includes/admin/meta-boxes/views/html-product-data-advanced.php:
```diff
@@ -52,6 +54,22 @@
+				'label'       => __( 'Available for POS', 'woocommerce' ),
+				'desc_tip'    => true,
+				'description' => __( 'Controls whether this product appears in the Point of Sale system.', 'woocommerce' ),
+			)
```
What do you think about using the `unchecked_value` parameter so that, when saving, the value doesn't get accidentally changed when this isn't posted back?  

It hasn't been around long so not many places use it, but it should avoid accidents where the value ends up being saved as if it is false because some integration didn't know to include value in the save request.

https://github.com/woocommerce/woocommerce/blob/8ddf51baae56b526c505c21902d4312d4db33cc7/plugins/woocommerce/includes/admin/wc-meta-box-functions.php#L176-L181

malinajirka replied:
Thanks for the suggestion @prettyboymp ! I'll try to implement it in a follow up PR to unblock this PR - I'd like to have it merged in 10.5.0. Considering this improvement is not implemented for tons of other fields, I assume it is nota blocker for the release.

prettyboymp reviewed (approved):
Everything looked good from my end.  Just one small suggestion.

malinajirka commented on plugins/woocommerce/src/Internal/ProductFeed/Integrations/POSCatalog/POSIntegration.php:
```diff
@@ -54,7 +54,16 @@ public function get_id(): string {
-			'type' => array( 'simple', 'variable', 'variation' ),
+			'type'      => array( 'simple', 'variable', 'variation' ),
+			// phpcs:ignore WordPress.DB.SlowDBQuery.slow_db_query_tax_query
+			'tax_query' => array(
```
⚠️ I'll need to continue testing this. I'm pretty sure something is broken - likely this part. The catalog file contains less than 800 items on my store even though it should contain around 870-880.

malinajirka replied:
okay, so this was partially a false alarm - the filtering works correctly. I just at some point marked one variable product with 75 variations as hidden in POS and the code correctly filtered them out. 

However, the `total` field in `/catalog` response is off. It excludes products that are marked as hidden, but doesn't exclude variations. We filter variations out in the FeedValidator (this was discussed [here](https://github.com/woocommerce/woocommerce/pull/62534#discussion_r2667709922)) which is processed after the `total` count has already been set with wc_get_products. 

The question is how to go about it. 

1. We could update the way we count the `total`, but it'd require 2-3 sql queries. 
2. We could add `pos-hidden` term to all variations of variable products.
3. We could ignore the fact that `total` is off in the /catalog endpoint, since the actual data are correct. My main worry here is that it'll be more difficult to spot bugs in the generation and debug customer issues.

malinajirka replied:
I've implemented 2 and I quite like the solution. I'll spend a bit more time testing it and ask for re-review when I'm confident it works as expected.

malinajirka committed:
Sync pos-hidden taxonomy to variations for accurate catalog totals

When a variable product is hidden from POS, its variations should also be
  excluded from the catalog feed and counted correctly in the total.

  Previously, the pos_product_visibility taxonomy was only applied to parent
  products, and variations were filtered out during feed generation via
  FeedValidator. This caused the catalog endpoint to return incorrect totals
  since variations were counted in the query but filtered out later.

  This change:
  - Registers pos_product_visibility taxonomy for product_variation post type
  - Introduces POSProductVisibilitySync class to centralize visibility logic
  - Syncs pos-hidden term to all variations when parent visibility changes
  - Inherits pos-hidden term when new variations are created
  - Removes FeedValidator variation check (no longer needed)

  Now the tax_query correctly filters both products and variations at query time,
  ensuring accurate totals in the catalog API response.

malinajirka committed:
Add pos_products_only filter to variations REST API endpoint

Add the pos_products_only parameter to the /variations and /products/<id>/variations endpoints to filter out variations with the pos-hidden term, matching the existing implementation in the products endpoint.

Changes:
- Add pos_products_only boolean parameter to get_collection_params()
- Add tax_query filter in prepare_objects_query() to exclude pos-hidden variations
- Add tests for the new filter behavior

This enables POS clients to fetch only variations visible in Point of Sale when querying the variations endpoint with pos_products_only=true.

malinajirka committed:
Guard POS visibility save logic with feature flag check

When the POS feature is disabled, the _visible_in_pos checkbox is not rendered in the admin product edit UI. However, the save logic was executing unconditionally. This caused $_POST['_visible_in_pos'] to be unset, resulting in $visible_in_pos becoming false and unintentionally hiding products from POS.

This fix wraps the POS visibility save logic with a feature flag check, matching the UI logic that also checks feature_is_enabled('point_of_sale') before rendering the checkbox. This prevents data loss when editing products while the POS feature is disabled.

malinajirka committed:
Fix lint issues in class-wc-rest-product-variations-controller-tests.php

malinajirka commented:
Thanks for bearing with me @RadoslavGeorgiev! 

I found out that the solution with filtering out variations within the FeedValidator had a side-effect - the `total` field in the response of `/catalog` endpoint was off, since totals were being calculated before the filtering. 

Considering this issue and considering that the `/variations` endpoint couldn't support `pos_products_only` parameter, I decided to enable the taxonomy for variations. I introduced `pos_products_only` parameter for `/wc/v3/variations` endpoint and I introduced POSProductVisibilitySync class that takes care of syncing variations with its parent. I have updated the test scenarios - I added testing for `/variations` endpoint. 

With that, I think this is hopefully ready for final review of the recent commits. I'm not aware of any downsides or bugs or open questions anymore.

Thank you and let me know what you think about the recent changes.

RadoslavGeorgiev reviewed (approved):
Everything looks very well!

I tested the UI & updates, and they work as expected. The one thing I'd suggest is to add unit tests to `POSProductVisibilitySync` :)

malinajirka committed:
Add tests for POSProductVisibilitySync.php
